File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys
input = sys.stdin.readline

n = int(input())
stack = []
current = 1
res = []
possible = True

for _ in range(n):
    a = int(input())

    while current <= a:
        stack.append(current)
        res.append("+")
        current += 1

    if stack[-1] == a:
        stack.pop()
        res.append("-")
    else:
        possible = False
        break

if possible:
    for op in res:
        print(op)
else:
    print("NO")

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringBuilder sb = new StringBuilder();
    Stack<Integer> stack = new Stack<>();
    int n, current;
    boolean flag;

    void solution() throws Exception {
        n = Integer.parseInt(br.readLine());
        current = 1;

        for (int i = 0; i < n; i++) {
            int k = Integer.parseInt(br.readLine());

            while (k >= current) {
                stack.push(current);
                sb.append("+").append("\n");
                current++;
            }

            flag = false;
            while (!stack.empty()) {
                int p = stack.pop();
                sb.append("-").append("\n");

                if (p == k) {
                    flag = true;
                    break;
                }
            }

            if (!flag) {
                System.out.println("NO");
                return;
            }
        }

        System.out.println(sb);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}